contracts/escrow_fee_distributor/src/contract.rs

```diff
@@ -0,0 +1,646 @@
+        ExecuteMsg::UpdateConfig {
+            max_limit_accounts_of_claim,
+            checkpoint_token_enabled,
+        } => update_config(
+            deps,
+            info,
+            max_limit_accounts_of_claim,
+            checkpoint_token_enabled,
+        ),
+        ExecuteMsg::Receive(msg) => receive_cw20(deps, env, info, msg),
+    }
+}
+
+/// ## Description
+/// Receives a message of type [`Cw20ReceiveMsg`] and processes it depending on the received template.
+/// If the template is not found in the received message, then an [`ContractError`] is returned,
+/// otherwise returns the [`Response`] with the specified attributes if the operation was successful
+fn receive_cw20(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    cw20_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    match from_binary(&cw20_msg.msg)? {
+        Cw20HookMsg::Burn {} => {
+            let mut config: Config = CONFIG.load(deps.storage)?;
+            if info.sender != config.astro_token {
+                return Err(ContractError::Unauthorized {});
+            }
+            if config.checkpoint_token_enabled
+                && (env.block.time.seconds() > config.last_token_time + TOKEN_CHECKPOINT_DEADLINE)
+            {
+                calc_checkpoint_token(deps.branch(), env, &mut config)?;
+                CONFIG.save(deps.storage, &config)?;
+            }
+
+            Ok(Response::new())
+        }
+    }
+}
+
+/// ## Description
+/// Update the token checkpoint. Returns the [`Response`] with the specified attributes if the
+/// operation was successful, otherwise returns the [`ContractError`].
+fn checkpoint_token(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let mut config: Config = CONFIG.load(deps.storage)?;
+
+    if info.sender != config.owner
+        && (!config.checkpoint_token_enabled
+            || env.block.time.seconds() < (config.last_token_time + TOKEN_CHECKPOINT_DEADLINE))
+    {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    calc_checkpoint_token(deps.branch(), env, &mut config)?;
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new().add_attributes(vec![attr("action", "checkpoint_token")]))
+}
+
+/// ## Description
+/// Calculates the total number of tokens to be distributed in a given week.
+fn calc_checkpoint_token(mut deps: DepsMut, env: Env, config: &mut Config) -> StdResult<()> {
+    let distributor_balance = query_token_balance(
+        &deps.querier,
+        config.astro_token.clone(),
+        env.contract.address.clone(),
+    )?;
+
+    let to_distribute = distributor_balance.checked_sub(config.token_last_balance)?;
+    let mut last_token_time = config.last_token_time;
+
+    let since_last = env.block.time.seconds() - last_token_time;
+
+    config.last_token_time = env.block.time.seconds();
+
+    let mut current_week = last_token_time / WEEK * WEEK; // week alignment
```

Comment:
We need to use timestamp to calculate amount of tokens per week, because we can run checkpoint everyday. But, for saving total amount tokens per week we can use period